### PR TITLE
[RFC] rfstool:  scripted rootfs builds. 

### PR DESCRIPTION
**Pull request by tnotw**
## Reason for This PR

#1732 Outlines one possible approach to building rootfs filesystems.  

## Description of Changes

Adds the "rfstool" and associated Dockerfile for building a container image with wget and libguestfs-tools installed.   The tool is currently restricted to build just the minimal test rootfs.  It will fetch a minifs image from the Alpine repos, extract it, then customise the OS in order to make it suitable for running on Firecracker.

To test:

1. Build Docker image from the Dockerfile in the commit
2. Amend the line `RFSCTR_IMAGE=` in the rfstool script to point to your new docker image
3. Run `./tools/rfstool build` 

Expected output: 
```
$ ./tools/rfstool build
[Firecracker rfstool] Starting rootfs minimal build ...
[   0.0] Examining the guest ...
[  40.4] Setting a random seed
virt-customize: warning: random seed could not be set for this type of 
guest
[  40.9] Installing packages: ca-certificates
[  50.6] Installing packages: curl
[  62.1] Installing packages: iperf3
[  73.1] Installing packages: iproute2
[  87.4] Installing packages: iptables
[  99.4] Installing packages: openrc
[ 112.3] Installing packages: openssh-server
[ 125.7] Installing packages: util-linux
[ 148.8] Linking: /etc/init.d/agetty.ttyS0 -> agetty
[ 148.9] Running: echo ttyS0 >/etc/securetty
[ 149.9] Running: rc-update add agetty.ttyS0 default
[ 151.0] Running: rc-update add devfs boot
[ 152.2] Running: rc-update add procfs boot
[ 153.4] Running: rc-update add sysfs boot
[ 154.6] Running: rc-update add local default
[ 156.0] Writing: /etc/local.d/umount.stop
[ 156.1] Changing permissions of /etc/local.d/umount.stop to 0700
[ 156.3] Setting passwords
[ 170.1] Finishing off
[Firecracker rfstool] Minimal rootfs built and available at: ./firecracker/build/img/local/x86_64/minimal/fsfiles/boottime-rootfs.ext4
```

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

**Comment by acatangiu** [comment]
Hi @tnotw, thanks for the contribution! Over time we've had a few requests to provide an automated rootfs generation script, so we're excited to see this PR.

We'll try to do an in-depth review some time this week.

**Comment by tnotw** [comment]
Hi @iggy thanks for your comments - I'll review the Alpine docs.  Perhaps this isn't quite working as I think......  I'll park making any amendments to your other comments until after project feedback.

File: tools/devtool [hunk 1]
```diff
@@ -771,6 +783,85 @@ cmd_tag() {
...
+
+    # make fsfiles
+    mkdir -p "$img_local/minimal/fsfiles" \
+      || die "Cannot create $img_local/fsfiles. Aborting"
+
+    say "Creating rootfs file..."
+    ( cd "$img_local/minimal/fsfiles" \
+      && virt-make-fs \
+           --size=30M \
+           --type=ext4 \
```
[the remark this answers is not in the document]
Comment (tnotw) [reply tools/devtool]: I was just really trying to replicate the `hello-rootfs.ext4` rootfs image

**Comment by tnotw** [comment]
Something along these lines perhaps?  A clone and cut down of the devtool with `virt-make-fs` and `virt-customize` being run in the container.  There is a `tnotw/img:v01` container available in Docker should anyone wish to give it a spin, or the Dockerfile is in the commit.  Only an x86_64 container currently.

File: tools/rfstool [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+RFS_ALPINE_X86_64_SHA256="ad753d802048fa902e4d8b35cc53656de8ed0e6d082246089a"
+RFS_ALPINE_X86_64_SHA256+="11a86014b0f1a5"
+
+# Rootfs Alpine aarch64 package name, url location, and sha256.
+RFS_ALPINE_AARCH64_GZ="alpine-minirootfs-${RFS_ALPINE_VERS}-aarch64.tar.gz"
+RFS_ALPINE_AARCH64_URL="${RFS_ALPINE_ORIGIN}"
+RFS_ALPINE_AARCH64_URL+="/alpine/v${RFS_ALPINE_VERS%.*}/releases/aarch64/"
+RFS_ALPINE_AARCH64_URL+="/${RFS_ALPINE_AARCH64_GZ}"
+RFS_ALPINE_AARCH64_SHA256="425be487ecc4bd0333fd84ec8c6c4b971a32446be76df26e7"
+RFS_ALPINE_AARCH64_SHA256+="add04be95987fa8"
```
Comment (serban300) [inline tools/rfstool]: Can we store this in an associative array or some sort of key-value object for readability ?  Again I'm not an expert in bash so I'm not sure about this. Ideally we could have something like (please ignore the syntax. It's inspired by javascript, but I hope we can do something like this in bash):
```
images = {
  minimal: {
    x86_64: {
      gz: ...
      url: ....
      sha: ....
    }
  }
}
```

This way we can then have some generic logic for downloading the image. If you know the image name (e.g. minimal`) and architecture (e.g. x86_64). You can just look in the key-value store and find all the parametrs that you need in order to download the sources.

**Comment by tnotw** [comment]
Hi @serban300 

Thank you for your review and feedback.
Apologies, this hasn't been a very coherent attempt at contributing on my part.

The fix I was proposing to address #1564 is the addition of these commands to the Alpine rootfs images: 

```
rc-update add local default
echo "umount /" >/etc/local.d/umount.stop
chmod 700 /etc/local.d/umount.stop
```

Where the intention is that when the "local" daemon runs on shutdown it unmounts and hence syncs the root disk, hopefully eradicating the corruption mentioned in the ticket.  I'm not really sure if that fixes the issue but anyway......  I had nowhere to add that in the codebase and no way of modifying the rootfs images,  so the code in this branch is really a misguided attempt at "sharing" a solution,  i.e. a rootfs build with the changes applied.  In retrospect, I should have just outlined the above on the ticket and asked if that sounded sensible.  Apologies again. 

Thanks very much for your code review - agree with all the points.

**Comment by iulianbarbu** [comment]
Hey, @tnotw ! As @serban300 said, having a tool that builds rootfs images brings a lot of value. Closing the PR it was not something that we had in mind. We can still push forward with this PR, but it will need some updates. It has to be linked to an issue which explains the nature of the contribution. A simple "rootfs build tool proposal" issue might be enough. What do you think?

**Comment by dianpopa** [comment]
> Hey, @tnotw ! As @serban300 said, having a tool that builds rootfs images brings a lot of value. Closing the PR it was not something that we had in mind. We can still push forward with this PR, but it will need some updates. It has to be linked to an issue which explains the nature of the contribution. A simple "rootfs build tool proposal" issue might be enough. What do you think?

There are some useful info as far as rootfs creation tool [here](https://github.com/firecracker-microvm/firecracker/issues/1418).

**Comment by tnotw** [comment]
Ok, apologies, I wasn't completely convinced this wasn't just noise for you,  but I'll have another go at it...

File: tools/rfstool (hunk 2 above)
Comment (tnotw) [reply tools/rfstool]: If the Alpine version is expected to be the same on both architectures,  then I think this just becomes:
```
RFS_ALPINE_GZ="alpine-minirootfs-${RFS_ALPINE_VERS}-${HOSTTYPE}.tar.gz"
RFS_ALPINE_URL="${RFS_ALPINE_ORIGIN}"
RFS_ALPINE_URL+="/alpine/v${RFS_ALPINE_VERS%.*}/releases/${HOSTTYPE}/"
RFS_ALPINE_URL+="/${RFS_ALPINE_GZ}"
```
And GZ and URL don't need to be included in any associative array,  just the SHA value.  I think.  

Currently however, the Alpine version for "minimal" differs between architecture, so the rootfs stored in S3 seem to me to be 3.8 on x86_64, 3.9 on aarch64.  So I guess it depends on the intention.  If it's to track each rootfs version per architecture,  i.e. try and match the discrepancies (which I appreciate could be intentional), then yes,  I will put in an array as you suggest.  If it's to clobber anything that's currently there and keep test artefacts aligned between architectures, then perhaps I don't need to.

File: tools/rfstool [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+    for dir in "$FC_BUILD_DIR"; do
+        mkdir -p "$dir" || die "Error: cannot create dir $dir"
+        [ -x "$dir" ] && [ -w "$dir" ] || \
+            {
+                say "Wrong permissions for $dir. Attempting to fix them ..."
+                chmod +x+w "$dir"
+            } || \
+            die "Error: wrong permissions for $dir. Should be +x+w"
+    done
+}
```
[the remark this answers is not in the document]
Comment (tnotw) [reply tools/rfstool]: Yup,  moved to a shared "functions" file.

File: tools/rfstool [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+
+    cmd=cmd_$1
+    shift
+
+    # $@ is now a list of command-specific args
+    #
+    $cmd "$@"
+}
+
+main "$@"
```
[the remark this answers is not in the document]
Comment (tnotw) [reply tools/rfstool]: Yup,  moved to a shared "functions" file.

File: tools/rfstool (hunk 2 above)
Comment (tnotw) [reply tools/rfstool]: Also,  I'm not sure that bash is very good at associative arrays, from what I've been reading,  bash's associative arrays are "one-dimensional", i.e. don't nest well.   Unless I've just not found the right  way....

**Comment by iulianbarbu** [comment]
> > Hey, @tnotw ! As @serban300 said, having a tool that builds rootfs images brings a lot of value. Closing the PR it was not something that we had in mind. We can still push forward with this PR, but it will need some updates. It has to be linked to an issue which explains the nature of the contribution. A simple "rootfs build tool proposal" issue might be enough. What do you think?
> 
> There are some useful info as far as rootfs creation tool [here](https://github.com/firecracker-microvm/firecracker/issues/1418).

I think this PR is a step which precedes the work need for #1418 . It is related, nevertheless.

**Comment by tnotw** [comment]
Hi @iulianbarbu,  Sure.  I've pushed what I have so far,  I can see about adding something to address #1418

**Comment by tnotw** [comment]
So for a 100G, ext2 filesystem, Alpine image, with the Stream binary installed say, you would do something like this: 
```
+# "ext2_100G_stream" profile details
+declare -A RFS_STREAM=(
+    [NAME]="stream.rootfs.ext2"
+    [SIZE]="100G"
+    [FS]="ext2"
+    [CUST]="--install openrc \
+           --copy-in /firecracker/stream:/bin/ \
+           --chmod 0755:/bin/stream \
+           --link agetty:/etc/init.d/agetty.ttyS0 \
+           --run-command 'echo ttyS0 >/etc/securetty' \
+           --run-command 'rc-update add agetty.ttyS0 default'"
+)
+
+
 # Global options received by $0
 # These options are not command-specific, so we store them as global vars
 OPT_UNATTENDED=false
@@ -224,6 +238,7 @@ cmd_build() {
         case "$1" in
             "-h"|"--help")      { cmd_help; exit 1;  } ;;
             "minimal")          { profile="minimal"; } ;;
+            "stream")           { profile="stream"; } ;;
             "ubuntu")           { profile="ubuntu";  } ;;
             "ubuntu_with_ssh")  { profile="ubuntu_with_ssh";  } ;;
             *)
@@ -249,6 +264,7 @@ cmd_build() {
     # The rootfs build is a combination of OS + profile
     case "${profile}" in
         "minimal")          { build_rootfs minimal alpine; } ;;
+        "stream")           { build_rootfs stream alpine; } ;;
         "ubuntu")           { build_rootfs ubuntu ubuntu_bionic; } ;;
         "ubuntu_with_ssh")  { build_rootfs ubuntu_with_ssh ubuntu_bionic; } ;;
         *)
```
So it is sortof data driven,  and there is sortof a separation between the base OS install and customization steps.  So all you essentially do is add a profile to the alpine base install.  But it's obviously not separated out into json, it's a bit clunky, and as you say, there are lots of limitations you hit fairly quickly.

So yeah, postpone or just bin.

**Pull request closed without merging**